[css-anchor-position-1] Remember last-successful position option at ResizeObserver delivery time

tuankiet65 opened the pull request:
#### 755cd64e24a56c80bf33dca159d61c94519fce32
<pre>
[css-anchor-position-1] Remember last-successful position option at ResizeObserver delivery time
<a href="https://rdar.apple.com/159225250">rdar://159225250</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297932">https://bugs.webkit.org/show_bug.cgi?id=297932</a>

Reviewed by Antti Koivisto.

RenderStyle::lastSuccessfulPositionTryFallbackIndex is _actually_ the index of the position
option being applied to the style. Because we stop trying as soon as we find a working option,
given a box&apos;s style after style resolution, lastSuccessfulPositionTryFallbackIndex is _also_
the index of the last successful position option. This implies that the last successful position
option is remembered at style resolution time. However, the spec [1] dictates that the
last successful position option is remember _when_ ResizeObserver events are dispatched.
This patch aligns with that behavior by separating trying position options into two phases:

* During style resolution, each option is tried, and RenderStyle::usedPositionOptionIndex
(renamed from ::lastSuccessfulPositionTryFallbackIndex) is set for each style to denote
what&apos;s the option used to generate the style.
* When ResizeObserver events are dispatched (after style resolution), we loop through
each position-try box and save the usedPositionOptionIndex of each box. This is the
last successful position option of the box.
* At the next style resolution, we use the saved last successful position option to
adjust which option should be tried first.

[1]: <a href="https://drafts.csswg.org/css-anchor-position-1/#last-successful-position-option">https://drafts.csswg.org/css-anchor-position-1/#last-successful-position-option</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-011-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/base-style-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateResizeObservations):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::unregisterPositionTryBox):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::usedPositionOptionIndex const):
(WebCore::RenderStyle::setUsedPositionOptionIndex):
(WebCore::RenderStyle::lastSuccessfulPositionTryFallbackIndex const): Deleted.
(WebCore::RenderStyle::setLastSuccessfulPositionTryFallbackIndex): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::recordLastSuccessfulPositionOptions):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::lastSuccessfulPositionOptionIndexFor):
(WebCore::Style::Scope::setLastSuccessfulPositionOptionIndexMap):
(WebCore::Style::Scope::forgetLastSuccessfulPositionOptionIndex):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::shouldInvalidateLastSuccessfulPositionOptionIndex):
(WebCore::Style::TreeResolver::styleForStyleable):
(WebCore::Style::TreeResolver::resolvePseudoElement):
(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):
(WebCore::Style::TreeResolver::PositionOptions::currentOption const):
(WebCore::Style::TreeResolver::sortPositionOptionsIfNeeded):
(WebCore::Style::TreeResolver::tryChoosePositionOption):
(WebCore::Style::lastSuccessfulPositionTryFallbackIndex): Deleted.
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/300890@main">https://commits.webkit.org/300890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9be333ad084b93c2bb4875ed616f636a21cf1eab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124110 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130943 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76231 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f3c3c2d7-5f32-47c7-87d0-4233f3cdc6c6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125987 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44546 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52408 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94409 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62637 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/304ad99d-ffef-410f-aeef-76647724d066) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127064 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35493 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111018 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75001 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9c281c53-38dd-4472-936c-54acc57708a9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29182 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74423 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105231 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133613 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38897 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102882 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51424 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107237 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102689 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26148 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48047 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26291 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47924 "Hash 9be333ad for PR 50010 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50901 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56673 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50362 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53710 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52036 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d7e2775c89e36539951135315f6fd1db668e4c65. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50010

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/43fdec8a48904242ac6b0b5222c2d12a6396e924. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50010

tuankiet65 commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -215,15 +225,28 @@ ResolvedStyle TreeResolver::styleForStyleable(const Styleable& styleable, Resolu
 
     // Preserve the last successful fallback by propagating it from the old to new style.
-    style->setLastSuccessfulPositionTryFallbackIndex(lastSuccessfulPositionTryFallbackIndex(existingStyle, style.get()));
+    // style->setShouldUpdateLastSuccessfulOptionIndex(shouldUpdateLastSuccessfulPositionIndex(existingStyle, style.get()));
```
I've removed this.

anttijk commented on Source/WebCore/style/StyleScope.h:
```diff
@@ -267,6 +272,7 @@ class Scope final : public CanMakeWeakPtr<Scope>, public CanMakeCheckedPtr<Scope
         bool operator==(const AnchorPosition&) const = default;
     };
     SingleThreadWeakHashMap<const RenderBoxModelObject, AnchorPosition> m_anchorPositionsOnLastUpdate;
+    HashMap<AnchorPositionedKey, size_t> m_lastSuccessfulPositionOptionIndexes;
```
AnchorPositionedKey refs Element and I think so far it is only used for transient data structures. I'm not entirely sure this patch guarantees cleanup in all case (when elements are removed from there tree etc)? Otherwise we'd leak Elements.

tuankiet65 replied:
The intention here is to have a unique way to identify an element + optional pseudo-element. `AnchorPositionedKey` is the most straightforward way, but we can do better without ref-counting the `Element`. Maybe `Element::nodeIdentifier()` + `PseudoElementIdentifier`?

anttijk replied:
forgetLastSuccessfulPositionOptionIndex is called when position try box is unregistered so maybe that is sufficient to ensure cleanup. Maybe you can put `ASSERT(m_lastSuccessfulPositionOptionIndexes.isEmpty())` to Style::Scope destructor to prove that?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c657c62fc950db95fc06b3e6d6638466ad7dc530. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50010

tuankiet65 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10282,6 +10282,19 @@ void Document::updateResizeObservations(Page& page)
+            auto styleable = Styleable::fromRenderer(positionTryBox);
+            ASSERT(styleable);
+
+            if (auto usedPositionOptionIndex = positionTryBox.style().usedPositionOptionIndex())
+                m_styleScope->rememberLastSuccessfulPositionOptionIndex(*styleable, *usedPositionOptionIndex);
+            else
+                m_styleScope->forgetLastSuccessfulPositionOptionIndex(*styleable);
+        }
+    }
```
I changed it to your approach.

tuankiet65 committed:
[css-anchor-position-1] Remember last-successful position option at ResizeObserver delivery time
rdar://159225250
https://bugs.webkit.org/show_bug.cgi?id=297932

Reviewed by Antti Koivisto.

RenderStyle::lastSuccessfulPositionTryFallbackIndex is _actually_ the index of the position
option being applied to the style. Because we stop trying as soon as we find a working option,
given a box's style after style resolution, lastSuccessfulPositionTryFallbackIndex is _also_
the index of the last successful position option. This implies that the last successful position
option is remembered at style resolution time. However, the spec [1] dictates that the
last successful position option is remember _when_ ResizeObserver events are dispatched.
This patch aligns with that behavior by separating trying position options into two phases:

* During style resolution, each option is tried, and RenderStyle::usedPositionOptionIndex
(renamed from ::lastSuccessfulPositionTryFallbackIndex) is set for each style to denote
what's the option used to generate the style.
* When ResizeObserver events are dispatched (after style resolution), we loop through
each position-try box and save the usedPositionOptionIndex of each box. This is the
last successful position option of the box.
* At the next style resolution, we use the saved last successful position option to
adjust which option should be tried first.

[1]: https://drafts.csswg.org/css-anchor-position-1/#last-successful-position-option

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-011-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/base-style-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateResizeObservations):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::unregisterPositionTryBox):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::usedPositionOptionIndex const):
(WebCore::RenderStyle::setUsedPositionOptionIndex):
(WebCore::RenderStyle::lastSuccessfulPositionTryFallbackIndex const): Deleted.
(WebCore::RenderStyle::setLastSuccessfulPositionTryFallbackIndex): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::recordLastSuccessfulPositionOptions):
* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::lastSuccessfulPositionOptionIndexFor):
(WebCore::Style::Scope::setLastSuccessfulPositionOptionIndexMap):
(WebCore::Style::Scope::forgetLastSuccessfulPositionOptionIndex):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::shouldInvalidateLastSuccessfulPositionOptionIndex):
(WebCore::Style::TreeResolver::styleForStyleable):
(WebCore::Style::TreeResolver::resolvePseudoElement):
(WebCore::Style::TreeResolver::generatePositionOptionsIfNeeded):
(WebCore::Style::TreeResolver::PositionOptions::currentOption const):
(WebCore::Style::TreeResolver::sortPositionOptionsIfNeeded):
(WebCore::Style::TreeResolver::tryChoosePositionOption):
(WebCore::Style::lastSuccessfulPositionTryFallbackIndex): Deleted.
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/300890@main
